spec/helpers/tcp_gate_scheduler.rb

```diff
@@ -0,0 +1,214 @@
+# frozen_string_literal: true
+
+# This is a special scheduler for testing compatibility to Fiber.scheduler of functions using a TCP connection.
+#
+# It works as a gate between the client and the server.
+# Data is transferred only, when the scheduler receives wait_io requests.
+# The TCP communication in a C extension can be verified in a (mostly) timing insensitive way.
+# If a call does IO but doesn't call the scheduler, the test will block and can be caught by an external timeout.
+#
+#   PG.connect
+#    port:5444                    TcpGateScheduler                     DB
+#  -------------      ----------------------------------------      --------
+#  | scheduler |      | TCPServer                  TCPSocket |      |      |
+#  |   specs   |----->|  port 5444                  port 5432|----->|Server|
+#  -------------  ^   |                                      |      | port |
+#                 '-------  wait_readable:  <-send data--    |      | 5432 |
+#           observe fd|     wait_writable:  --send data->    |      --------
+#                     ----------------------------------------
+
+module Helpers
+class TcpGateScheduler < Scheduler
```

Comment:
I guess you cannot know for certain that some code does or does not block. But we do have one "metric" for blocking, `rb_nogvl`. But this isn't reliable, even if it can be useful.

https://github.com/ruby/ruby/pull/4779

If we are comfortable with assuming "blocking" = "release GVL", then this will catch those cases.